File: gamepack/Decompiled/NodeHashTable.java
public final class NodeHashTable {
   int af;
   int au = 0;
   Node ac;
   Node aw;
   Node[] an;

   public NodeHashTable(int var1) {
      this.af = var1;
      this.an = new Node[var1];

      for(int var2 = 0; var2 < var1; ++var2) {
         Node var3 = this.an[var2] = new Node();
         var3.hc = var3;
         var3.hg = var3;
      }

   }

   public Node get(long var1) {
      Node var3 = this.an[(int)(var1 & (long)(this.af - 1))];

      for(this.aw = var3.hc; this.aw != var3; this.aw = this.aw.hc) {
         if (this.aw.hr == var1) {
            Node var4 = this.aw;
            this.aw = this.aw.hc;
            return var4;
         }
      }

      this.aw = null;
      return null;
   }

   public void put(Node var1, long var2) {
      if (var1.hg != null) {
         var1.remove();
      }

      Node var4 = this.an[(int)(var2 & (long)(this.af - 1))];
      var1.hg = var4.hg;
      var1.hc = var4;
      var1.hg.hc = var1;
      var1.hc.hg = var1;
      var1.hr = var2;
   }

   public Node first() {
      this.au = 0;
      return this.next();
   }

   public Node next() {
      Node var1;
      if (this.au > 0 && this.ac != this.an[this.au - 1]) {
         var1 = this.ac;
         this.ac = var1.hc;
         return var1;
      } else {
         do {
            if (this.au >= this.af) {
               return null;
            }

            var1 = this.an[this.au++].hc;
         } while(var1 == this.an[this.au - 1]);

         this.ac = var1.hc;
         return var1;
      }
   }
}

File: find-names/src/main/java/mapper/identifiers/classes/NodeHashTable.kt
package mapper.identifiers.classes

import mapper.abstractclasses.IdentityMapper
import mapper.abstractclasses.OrderMapper
import mapper.annotations.DependsOn
import mapper.annotations.MethodParameters
import mapper.predicateutilities.and
import mapper.predicateutilities.predicateOf
import mapper.predicateutilities.type
import mapper.predicateutilities.withDimensions
import mapper.wrappers.ClassWrapper
import mapper.wrappers.FieldWrapper
import mapper.wrappers.InstructionMapper
import mapper.wrappers.MethodWrapper
import org.objectweb.asm.Opcodes.*
import org.objectweb.asm.Type.*
import org.runestar.client.common.startsWith

@DependsOn(Node::class)
class NodeHashTable : IdentityMapper.Class() {

    override val predicate = predicateOf<ClassWrapper> { it.superType == Any::class.type }
            .and { it.instanceFields.count { it.type == type<Node>().withDimensions(1) } == 1 }
            .and { it.instanceFields.count { it.type == type<Node>() } == 2 }
            .and { it.interfaces.isEmpty() }

    @DependsOn(Node::class)
    class buckets : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<FieldWrapper> { it.type == type<Node>().withDimensions(1) }
    }

    class index : OrderMapper.InConstructor.Field(NodeHashTable::class, 0) {
        override val predicate = predicateOf<InstructionMapper> { it.opcode == PUTFIELD && it.fieldType == INT_TYPE }
    }

    class size : OrderMapper.InConstructor.Field(NodeHashTable::class, 1) {
        override val predicate = predicateOf<InstructionMapper> { it.opcode == PUTFIELD && it.fieldType == INT_TYPE }
    }

    @MethodParameters("key")
    @DependsOn(Node::class)
    class get : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<Node>() }
                .and { it.arguments.startsWith(LONG_TYPE) }
                .and { it.arguments.size in 1..2 }
    }

    @MethodParameters("node", "key")
    @DependsOn(Node::class)
    class put : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == VOID_TYPE }
                .and { it.arguments.startsWith(type<Node>(), LONG_TYPE) }
                .and { it.arguments.size in 2..3 }
    }

    @MethodParameters()
    @DependsOn(Node::class)
    class first : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<Node>() }
                .and { it.arguments.size in 0..1 }
                .and { it.instructions.none { it.opcode == ICONST_1 } }
    }

    @MethodParameters()
    @DependsOn(Node::class, get::class)
    class next : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<MethodWrapper> { it.returnType == type<Node>() }
                .and { it.arguments.size in 0..1 }
                .and { it.instructions.any { it.opcode == ICONST_1 } }
                .and { it != method<get>() }
    }

//    @MethodParameters()
//    class clear : IdentityMapper.InstanceMethod() {
//        override val predicate = predicateOf<Method2> { it.returnType == VOID_TYPE }
//                .and { it.arguments.size in 0..1 }
//    }

    @DependsOn(Node::class, next::class)
    class current : OrderMapper.InMethod.Field(next::class, 0) {
        override val predicate = predicateOf<InstructionMapper> { it.isField && it.fieldType == type<Node>() }
    }

    @DependsOn(current::class, Node::class)
    class currentGet : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<FieldWrapper> { it.type == type<Node>() && it != field<current>() }
    }
}